bundles/org.eclipse.ui.navigator.resources/src/org/eclipse/ui/internal/navigator/resources/actions/ResourceMgmtActionProvider.java

```diff
@@ -117,14 +116,15 @@ public void fillContextMenu(IMenuManager menu) {
 		boolean isProjectSelection = true;
 		boolean hasOpenProjects = false;
 		boolean hasClosedProjects = false;
-		boolean hasBuilder = true; // false if any project is closed or does not
-									// have builder
-
-		Iterator<IProject> projects = selectionToProjects(selection).iterator();
-
-		while (projects.hasNext() && (!hasOpenProjects || !hasClosedProjects || hasBuilder || isProjectSelection)) {
-			IProject project = projects.next();
+		boolean hasBuilder = true; // false if any project is closed or does not have builder
+		List<IProject> projects = selectionToProjects(selection);
+		boolean selectionContainsNonProject = projects.size() < selection.size();
 
+		for (IProject project : projects) {
+			if (hasOpenProjects && hasClosedProjects && !hasBuilder && !isProjectSelection) {
+				// we've set all booleans of interest; no need to loop any further
+				break;
+			}
 			if (project == null) {
```

Comment:
Can you explain why project can be null here and how that would affect "selectionContainsNonProject"?

Reply:
There's no way that this could be null, however given I don't know the code base well I was trying to touch as little as possible in my changes, so I left this as-is. 

From what I can see commit 18df0075dfd74d07ae9c384ede85e94233b36cdd from 2017 could have / should have removed this, but did not. Regardless: this is cleaned up in the commit added today